### ISSUE TITLE
operate first a pulp+python_plugin instance (or more)

### ISSUE THREAD

**Issue by goern** [operate-first/support#176]
```feature
Feature: op1st is operating one Pulp instance

Scenario: multi-index pulp
  Given we can deploy Pulp via an Operator and kustomize manifests 
  When pip installed a module from it
  And there are multiple variants of the same package version
  Then pip should separate the multiple variants by multiple index url

Scenario: use RBAC on multi-index pulp
  Given I have access to Pulp
  When I publish a module to an index URL
  And when I am not the 'owner' of that index
  Then I should be denied from publishing the module

Scenario: publish module
  Given I am a Tekton pipeline user
  When I publish a module to an index url
  Then I see the module on the simple index
  ```

@harshad16 @fridex @tumido this needs refinement

@fridex could you add the Pulp team?

**Comment by fridex** [comment]
> @fridex could you add the Pulp team?

CC @ipanova @fao89 @dralley

Feel free to add others you find relevant. As discussed in the meeting, we would like to deploy pulp with the pulp_python plugin.

**Comment by fao89** [comment]
adding @mikedep333 as he is the SME on https://github.com/pulp/pulp-operator

**Comment by fao89** [comment]
we currently provide 3 ways of installing pulp:
- https://github.com/pulp/pulp_installer
- https://github.com/pulp/pulp-oci-images
- https://github.com/pulp/pulp-operator

We have a brief explanation of them here:
https://pulpproject.org/installation-introduction/

**Comment by tumido** [comment]
Ohh, long time no see, Pulp team! :slightly_smiling_face: How are you doing these days? :slightly_smiling_face: Welcome! 

I think the `pulp-operator` serves our purpose the best. I think we'd like to be abstracted from the internals of Pulp as much as possible. If the experience of running an operator in active development in a prod-like environment would benefit the Pulp team, I see that as a plus as well.

**Comment by mikedep333** [comment]
Hi @tumido,

We would love for you to do that.

What internals do you see as important / remaining to be abstracted away?

**Comment by 4n4nd** [comment]
Hey @mikedep333, we do have one other operator (https://github.com/observatorium/operator) deployed which is in active development. We have set up the crds/clusterroles/bindings in a central location [here](https://github.com/operate-first/apps/tree/master/cluster-scope/base) and other required resources in a separate directory like [here](https://github.com/operate-first/apps/tree/master/observatorium/base). 
You should be able to follow the same structure for setting up the pulp-operator. If you have any suggestions or questions please lmk.

**Comment by tumido** [comment]
@mikedep333 
> What internals do you see as important / remaining to be abstracted away?

I don't think there's anything remaining to be abstracted away in the case of the operator. That's why I prefer it as the solution here. :slightly_smiling_face: I think we may get an idea on what might be improved once we start using it. Right now my comment was directed mostly to comparison of the 3 methods @fao89 outlined above - the operator is abstracting away tons of complexity compared to the other installers and is declarative. And we can appreciate that.

I'm gonna go ahead and start creating a namespace for the operator to live at - and we will automate this as a custom deployment of the operator (custom meaning directly deploying the `Deployment` resource, creating service account and so on) into this new namespace - similar to the observatorium operator @4n4nd linked above.

I'm also gonna create a new user group for you with full access to this new namespace so you can manage and monitor the operator yourself if you want.

The deployment of the operator will be managed via ArgoCD using the manifests copied/referenced from here:
https://github.com/pulp/pulp-operator/tree/main/deploy

Once the operator is available in the community operator hub we can either switch to a deployment from there or keep using a custom "manual" deployment for more rapid dev cycles on it if you want.

**Comment by fao89** [comment]
@tumido I'm a noob on k8s world, I never worked with ArgoCD. I have a "CI knowledge" of pulp-operator, meaning I only used pulp-operator on these cases: https://github.com/pulp/pulp-operator/actions/runs/728145612
But I see a great opportunity for us to improve our docs: https://pulp-operator.readthedocs.io/en/latest/
Let me know how can I help or at least what you are missing from the docs

**Comment by fridex** [comment]
Just a friendly ping here. What is the current state of this? We are monitoring this work on the package index meeting with Pulp team. Thanks in advance.

CC @ipanova

**Comment by tumido** [comment]
Yeah, sorry we had no upgrade on this so far, we've got hammered by a ton of work elsewhere. @fridex 

I see the operator didn't reach OperatorHub yet but you have CSV available. It also seems to me that the [cluster role](https://github.com/pulp/pulp-operator/blob/main/deploy/cluster_role.yaml)/[role](https://github.com/pulp/pulp-operator/blob/main/deploy/role.yaml) specified in the direct manifests is not yet prepared for an `AllNamespaces` role and if we deploy the operator this way the scope is limited to its current namespace only, is that a correct observation?

@fridex do you want to have the operator namespace scoped only within it's own namespace or available to multiple namespaces? I assume you'd rather to have the operator available globally, is that correct? If so, we either have to change the direct manifests a bit or create our own operator catalog source image and install via CSV.

**Comment by fridex** [comment]
> @fridex do you want to have the operator namespace scoped only within it's own namespace or available to multiple namespaces? I assume you'd rather to have the operator available globally, is that correct? If so, we either have to change the direct manifests a bit or create our own operator catalog source image and install via CSV.

Ideally, the operator could be available globally. Short-term, it would be great for us to have just one instance of pulp in one namespace for a selected group of people, small steps could work here. The very first outcome for us is the fact we can run pulp on op1st and can experiment with features it provides to us. The cluster-scope operator can be done in parallel (low priority for us now).

**Comment by tumido** [comment]
I'm sorry for the constant delays on this. I'm prioritizing this now I hope I can it something in place in few days.

**Comment by tumido** [comment]
Hey folks, so.. I can offer you 2 options. I think it's up to you to decide which way is more maintainable for you. Note - either of these solutions is temporary. Once you submit your operator to OperatorHub, this model changes - we would consume the operator manifest via subscription from community-operators.

### Option 1 - Direct manifests

Implemented in https://github.com/operate-first/apps/pull/663

Pulp team would need to track for changes all the  `CustomResourceDefinition`s, `ClusterRole`s.. basically any resource defined in that PR.

1. You would need to maintain it and update all the manifests defined within `cluster-scope/` path it in our repos. Basically copy and paste those resources back here if they change in your repos.
2. The resources within `pulp-operator/` path in that PR are transferable and can be deployed from any repo since they are namespace scoped and you already have full control over the `pulp-operator` namespace.

### Option 2 - Install via OLM via a custom catalog

Implemented in https://github.com/operate-first/apps/pull/664

This PR is based on your [`ClusterServiceVersion`](https://github.com/pulp/pulp-operator/tree/main/deploy/olm-catalog/pulp-operator) and defined a `CatalogSource`. Right now it points you my image, but the intention is that you own this image and keep the content of the catalog updated - Every time you change the CSV in your repos, you also update the catalog image. You can either use your own catalog or base it on [my catalog I've created for this purpose](https://github.com/tumido/olm-catalog).

My custom catalog is available for you, there's even [an updater script](https://github.com/tumido/olm-catalog/blob/main/scripts/update_pulp.sh) that will keep the catalog up to date with [`pulp-operator`](https://github.com/pulp/pulp-operator/) repository master branch. Once you push an updated catalog image, the rest of the update in cluster happens automatically.

This option is much easier to migrate once you submit your operator to OperatorHub since we would just point the `Subscription` resource to a different catalog.

### Summary

The decision is up to you, both approaches are valid. Either you want to maintain an OLM catalog for your dev purposes (you already have CSV up to date, so the overhead is not that big) or you'd rather copy and paste the cluster-scoped resources into our repository via PRs. Either is fine with us I think. :slightly_smiling_face: 

cc @fao89 @fridex @ipanova

**Comment by fao89** [comment]
we are planning to submit our operator to OperatorHub, so I would vote option 2

**Comment by tumido** [comment]
cc @HumairAK @4n4nd are we also good on using the custom catalog/subscription for the time being (until the pulp operator reaches OperatorHub)?

**Comment by 4n4nd** [comment]
yeah using the custom catalog/subscription sounds good to me :+1:

**Comment by tumido** [comment]
@fao89 @fridex 

Pulp operator is available at cluster scope. It's operated from the `pulp-operator` namespace which is owned by the `pulp` user group:

![image](https://user-images.githubusercontent.com/7453394/119329559-a469c700-bc85-11eb-9183-dcb0a5d52340.png)

Operator is up an running. @fridex can you please try deploying any `Pulp*` CR in any of your namespaces to see if it works?

Also.. a quick thought, @fridex do you want to have access to the `pulp-operator` namespace as well? (So you can access the operator logs or what not in case you need it..)

**Comment by fridex** [comment]
> @fao89 @fridex
> 
> Pulp operator is available at cluster scope. It's operated from the `pulp-operator` namespace which is owned by the `pulp` user group:

Awesome, thanks for the work 👍🏻 

> Operator is up an running. @fridex can you please try deploying any `Pulp*` CR in any of your namespaces to see if it works?

I tried to provision Pulp in thoth-test-core namespace. It looks like only postgres was provisioned:

![Screenshot_2021-05-24_11-59-22](https://user-images.githubusercontent.com/880687/119331348-88672500-bc87-11eb-8a4c-202f9b237902.png)

> Also.. a quick thought, @fridex do you want to have access to the `pulp-operator` namespace as well? (So you can access the operator logs or what not in case you need it..)

That might be good, but not essential as I do not have pulp expertise. Would it be possible to onboard Pulp team representatives @ipanova and/or @fao89?

**Comment by tumido** [comment]
@mikedep333 and @fao89, you already have access:
https://console-openshift-console.apps.zero.massopen.cloud/k8s/cluster/projects/pulp-operator

@ipanova wanna be added as well? :slightly_smiling_face:

**Comment by tumido** [comment]
@fridex here's the operator log for you, it seems to be failing due to quota on that namespace:
```
{\"ansible_loop_var\": \"item\", \"changed\": false, \"error\": 403, \"item\": \"redis\", \"msg\": \"Failed to create object: b'{\\\"kind\\\":\\\"Status\\\",\\\"apiVersion\\\":\\\"v1\\\",\\\"metadata\\\":{},\\\"status\\\":\\\"Failure\\\",\\\"message\\\":\\\"persistentvolumeclaims \\\\\\\\\\\"example-pulp-redis-data\\\\\\\\\\\" is forbidden: exceeded quota: thoth-test-core-custom, requested: requests.storage=1Gi, used: requests.storage=40Gi, limited: requests.storage=40Gi\\\",\\\"reason\\\":\\\"Forbidden\\\",\\\"details\\\":{\\\"name\\\":\\\"example-pulp-redis-data\\\",\\\"kind\\\":\\\"persistentvolumeclaims\\\"},\\\"code\\\":403}\\\\n'\", \"reason\": \"Forbidden\", \"status\": 403}\u001b[0m\n\r\nPLAY RECAP 
```

You probably want to request an increase by changing the manifest here:
https://github.com/operate-first/apps/blob/master/cluster-scope/base/core/namespaces/thoth-test-core/resourcequota.yaml

Full operator log attached.

[log.txt](https://github.com/operate-first/support/files/6531561/log.txt)

**Comment by fridex** [comment]
Thanks @tumido. Long-term, it would be probably better to separate Pulp. Would it be possible to create a separate namespace for pulp experiments so others involved have also access to it? Please do let me know if there is already a procedure for it.

**Comment by tumido** [comment]
@fridex sure, you can follow this doc here:
https://www.operate-first.cloud/users/support/docs/onboarding_to_cluster.md

You can open an issue in this repo using an onboarding template or DIY via a PR

For PR you can either use our onboarding script in the `apps` repo, as described in the doc linked above, or give a try to our [brand new cli tool](https://github.com/operate-first/opfcli/) if you feel adventurous).

1. create namespace resource and assign owners
```sh
cd operate-first/apps
# via script
scripts/onboarding.sh thoth-pulp-experiments thoth
# or via cli
opfcli create-project thoth-pulp-experiments thoth
```
2. set quota (this is still manual right now), please follow https://github.com/operate-first/apps/blob/master/docs/cluster-scope/add_resource_quotas.md
3. Import namespace to target cluster by editing `cluster-scoped/overlays/moc/zero/kustomization.yaml` and adding a new line to the `resources` list.

That should be all..

**Comment by fridex** [comment]
@tumido thanks for the manual - it looks like I would require gpg keys, @harshad16 was open to make this happen so I leave it to pros :) opened https://github.com/operate-first/support/issues/252 to track this

Thanks 👍🏻

**Comment by tumido** [comment]
@fridex your namespace should be available now at https://console-openshift-console.apps.zero.massopen.cloud/k8s/cluster/projects/thoth-pulp-experiments

LMK if it works properly now :slightly_smiling_face:

**Comment by fridex** [comment]
@tumido awesome, thanks! I was able to access the namespace and deploy pulp (partially). The operator started postgres and redis, but not pulp itself. Might be a resource limitation again? I see just 2 CPUs available in medium. Could you please check this? Thanks! 👍🏻 

![Screenshot_2021-05-26_13-26-29](https://user-images.githubusercontent.com/880687/119652335-3f010c00-be26-11eb-8179-fedffa998d1a.png)

**Comment by tumido** [comment]
Yeah, you're hitting quota again. You're creating a Pulp instance with 50Gi storage + redis allocates 1Gi + postgres allocates 8Gi -> 59Gi in total. Your quota is 40Gi now. I'll increase from 40Gi it to 60Gi for you on that namespace and we'll see where that goes. :slightly_smiling_face: 

```
*****************************************************\r\n\u001b[1;30mtask path: /opt/ansible/roles/pulp-api/tasks/main.yml:15\u001b[0m\n\u001b[0;36mskipping: [localhost] => {\"changed\": false, \"skip_reason\": \"Conditional result was False\"}\u001b[0m\n\r\nTASK [pulp-api : pulp-file-storage persistent volume claim] ********************\r\n\u001b[1;30mtask path: /opt/ansible/roles/pulp-api/tasks/main.yml:20\u001b[0m\n\u001b[0;35m[DEPRECATION WARNING]: evaluating 'file_storage' as a bare variable, this \u001b[0m\r\n\u001b[0;35mbehaviour will go away and you might need to add |bool to the expression in the\u001b[0m\r\n\u001b[0;35m future. Also see CONDITIONAL_BARE_VARS configuration toggle. This feature will\u001b[0m\r\n\u001b[0;35m be removed in version 2.12. Deprecation warnings can be disabled by setting \u001b[0m\r\n\u001b[0;35mdeprecation_warnings=False in ansible.cfg.\u001b[0m\n\u001b[0;31mfailed: [localhost] (item=pulp-file-storage) => {\"ansible_loop_var\": \"item\", \"changed\": false, \"error\": 403, \"item\": \"pulp-file-storage\", \"msg\": \"Failed to create object: b'{\\\"kind\\\":\\\"Status\\\",\\\"apiVersion\\\":\\\"v1\\\",\\\"metadata\\\":{},\\\"status\\\":\\\"Failure\\\",\\\"message\\\":\\\"persistentvolumeclaims \\\\\\\\\\\"example-pulp-file-storage\\\\\\\\\\\" is forbidden: exceeded quota: medium, requested: requests.storage=50Gi, used: requests.storage=9Gi, limited: requests.storage=40Gi\\\",\\\"reason\\\":\\\"Forbidden\\\",\\\"details\\\":{\\\"name\\\":\\\"example-pulp-file-storage\\\",\\\"kind\\\":\\\"persistentvolumeclaims\\\"},\\\"code\\\":403}\\\\n'\", \"reason\": \"Forbidden\", \"status\": 403}\u001b[0m\n\r\nPLAY RECAP 
```

**Comment by tumido** [comment]
Ok, now it seems to be complaining due to a different issue:

```
[pulp-api : Store admin password] *****************************************\r\n\u001b[1;30mtask path: /opt/ansible/roles/pulp-api/tasks/admin_password_configuration.yml:45\u001b[0m\n\u001b[0;32mok: [localhost] => {\"ansible_facts\": {\"admin_password\": \"Lh8NWJp2bUkqSHToZN28xw0Xzvm28ix4\"}, \"changed\": false}\u001b[0m\n\r\nTASK [pulp-api service] ********************************************************\r\n\u001b[1;30mtask path: /opt/ansible/roles/pulp-api/tasks/main.yml:77\u001b[0m\n\u001b[0;31mfailed: [localhost] (item=pulp-api) => {\"ansible_loop_var\": \"item\", \"changed\": false, \"error\": 422, \"item\": \"pulp-api\", \"msg\": \"Failed to create object: b'{\\\"kind\\\":\\\"Status\\\",\\\"apiVersion\\\":\\\"v1\\\",\\\"metadata\\\":{},\\\"status\\\":\\\"Failure\\\",\\\"message\\\":\\\"Service \\\\\\\\\\\"example-pulp-api-svc\\\\\\\\\\\" is invalid: spec.ports[0].nodePort: Invalid value: 24817: provided port is not in the valid range. The range of valid ports is 30000-32767\\\",\\\"reason\\\":\\\"Invalid\\\",\\\"details\\\":{\\\"name\\\":\\\"example-pulp-api-svc\\\",\\\"kind\\\":\\\"Service\\\",\\\"causes\\\":[{\\\"reason\\\":\\\"FieldValueInvalid\\\",\\\"message\\\":\\\"Invalid value: 24817: provided port is not in the valid range. The range of valid ports is 30000-32767\\\",\\\"field\\\":\\\"spec.ports[0].nodePort\\\"}]},\\\"code\\\":422}\\\\n'\", \"reason\": \"Unprocessable Entity\", \"status\": 422}\u001b[0m\n\r\nPLAY RECAP 
```

Operator is creating `example-pulp-api-svc` service and complains that `nodePort` has invalid value of 24817 (valid range is 30000-32767). Any idea where that comes from?

@fridex is deploying pulp manifest that looks like this:

```yaml
apiVersion: pulp.pulpproject.org/v1beta1
kind: Pulp
metadata:
  name: example-pulp
  namespace: thoth-pulp-experiments
spec:
  route_tls_termination_mechanism: Edge
  loadbalancer_port: 80
  image_pull_policy: IfNotPresent
  image_web: pulp-web
  file_storage:
    access_mode: ReadWriteMany
    size: 50Gi
  project: pulp
  tag: latest
  image: pulp
  loadbalancer_protocol: http
  registry: quay.io
  storage_type: File
```

any idea what's going on? @fao89 @ipanova

operator log attached:
[log.txt](https://github.com/operate-first/support/files/6548410/log.txt)

**Comment by fao89** [comment]
@tumido I'm not familiar with OCP, but we expand the nodeport range on minikube
https://github.com/pulp/pulp-operator/blob/main/.github/workflows/ci.yml#L37

**Comment by tumido** [comment]
Oh, sorry to hear that, I don't have a good news for you than. :disappointed:

In OCP this is more complicated. You need to ensure the port range is available and allowed in any underlying OCP provider (AWS, GCP, bare metal) and not blocked by any firewall on the infra level and above. Then you can [go and change the range](https://docs.openshift.com/container-platform/4.7/networking/configuring-node-port-service-range.html) for the whole cluster network.

Hm, I don't think you want your operator to be that much opinionated about how is the cluster set up and how the infrastructure beneath OCP behaves. In other words, you'd need to make this settings a prerequisite for Pulp, since this goes even beyond cluster-admin permissions.

**Comment by tumido** [comment]
I've noticed this `if` statement. Is there a way to workaround this and make it so it doesn't bind to a node port? (or am I missing something there..)

https://github.com/pulp/pulp-operator/blob/221c7652118d6c1c6dcda785fe5d651f14e0b101/roles/pulp-api/templates/pulp-api.service.yaml.j2#L25

**Comment by fao89** [comment]
> Oh, sorry to hear that, I don't have a good news for you than. 
> 
> In OCP this is more complicated. You need to ensure the port range is available and allowed in any underlying OCP provider (AWS, GCP, bare metal) and not blocked by any firewall on the infra level and above. Then you can [go and change the range](https://docs.openshift.com/container-platform/4.7/networking/configuring-node-port-service-range.html) for the whole cluster network.
> 
> Hm, I don't think you want your operator to be that much opinionated about how is the cluster set up and how the infrastructure beneath OCP behaves. In other words, you'd need to make this settings a prerequisite for Pulp, since this goes even beyond cluster-admin permissions.

@mikedep333 @dkliban @mdellweg ^



> I've noticed this `if` statement. Is there a way to workaround this and make it so it doesn't bind to a node port? (or am I missing something there..)
> 
> https://github.com/pulp/pulp-operator/blob/221c7652118d6c1c6dcda785fe5d651f14e0b101/roles/pulp-api/templates/pulp-api.service.yaml.j2#L25

I think we can do it, wdyt @chambridge ?

**Comment by tumido** [comment]
@fao89 here's the ticket https://pulp.plan.io/issues/8833

(In the end I had to create the Plan account I was refusing to last time, lol.. :smile: )

Btw. I've also noticed one more issue caused by the nodePort usage in this service (also described in the issue) - even if we bind a node port from the allowed range, it will still make Pulp API service basically a singleton for the cluster. NodePort means a "physical" port on the nodes, so it can be bound only to one service on the cluster. This results in no other Pulp resource anywhere on the cluster being able to deploy successfully if there's already a Pulp api server present.

**Comment by chambridge** [comment]
If you are deploying on OpenShift I'd suggest using
```
ingress_type: Route
route_tls_termination_mechanism: Edge
```
There was a recent PR that went in to fix the nodeport templating so its only used when that ingress_type is specified:
https://github.com/pulp/pulp-operator/pull/146

I think we could leverage https://pulp.plan.io/issues/8833 to make the "default" values in a more friendly range and updated the CRD so it could be configurable (so multiple nodeport pulp instances can be deployed)

**Comment by fao89** [comment]
@tumido with @chambridge help I was able to successfully deploy pulp, I tried to document it here:
https://github.com/pulp/pulp-operator/pull/149

**Comment by fridex** [comment]
I've just tried to provision pulp in `thoth-pulp-experiments` namespace, it looks like the issue reported above still persists (only postgres and redis are up). I see the referenced PR is merged, is there anything else blocking us from provisioning the instance? Thanks 👍🏻

**Comment by tumido** [comment]
@fridex 
I had to update the custom catalog with Pulp operator to get the update propagated to the cluster. The operator progressed and updated.

I think a better sollution in long run is to switch to the upstream catalog @fao89 mentioned in his docs. Once https://github.com/operate-first/apps/pull/703 merges we can reinstall the operator and you can try it again. And if Pulp team updates their catalog with an update/replacement CSV, the operator should progress automatically.

Using the official catalog, you should be able to follow @fao89's guide here: https://github.com/pulp/pulp-operator/blob/main/docs/quickstart.md

**Comment by tumido** [comment]
@fridex I'll also add you to the pulp-operator namespace, so you can observe the operator logs for yourself as well. :slightly_smiling_face:

**Comment by tumido** [comment]
@fridex we've now switched to the "official" catalog image via https://github.com/operate-first/apps/pull/703

I've reinstalled the operator on the cluster, so now it should be up to date.

You also have access to [`pulp-operator`](https://console-openshift-console.apps.zero.massopen.cloud/k8s/ns/pulp-operator/pods) namespace now.

**Comment by fridex** [comment]
Thanks 👍🏻 

I tried to provision pulp once again, but still no success:

From the operator logs I can see:
```
task path: /opt/ansible/roles/pulp-api/tasks/main.yml:77
failed: [localhost] (item=pulp-api) => {"ansible_loop_var": "item", "changed": false, "error": 422, "item": "pulp-api", "msg": "Failed to create object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"Service \\\\\"example-pulp-api-svc\\\\\" is invalid: spec.ports[0].nodePort: Invalid value: 24817: provided port is not in the valid range. The range of valid ports is 30000-32767\",\"reason\":\"Invalid\",\"details\":{\"name\":\"example-pulp-api-svc\",\"kind\":\"Service\",\"causes\":[{\"reason\":\"FieldValueInvalid\",\"message\":\"Invalid value: 24817: provided port is not in the valid range. The range of valid ports is 30000-32767\",\"field\":\"spec.ports[0].nodePort\"}]},\"code\":422}\\n'", "reason": "Unprocessable Entity", "status": 422}
``

**Comment by ipanova** [comment]
is not this related to this issue https://pulp.plan.io/issues/8833?
@chambridge @mikedep333  can you please keep an eye on this thread while @fao89 is on PTO? What needs to be done to progress with 8833?

**Comment by chambridge** [comment]
I'm happy to help jump on and debug if you want to give me access. I'm interested in the CR being used for the deploy as well as the logs. I wouldn't think you would be deploying where nodeport would be in use so I'm a bit confused how you're hitting the issue. 8833 could worked for this but really shouldn't be necessary for an OpenShift deployment.

**Comment by chambridge** [comment]
Can some one send me login information? I assume the above granted me login and RBAC privileges.

**Comment by 4n4nd** [comment]
@chambridge you shooould be able to use your @redhat.com account to log in

**Comment by tumido** [comment]
@chambridge Use https://console-openshift-console.apps.zero.massopen.cloud/ and select MOC SSO - then use your RH account via Google auth provider.

**Comment by chambridge** [comment]
The postgres & redis pods are currently stuck in pending waiting for PVCs to get created. Is anyone able to help with that?

![image](https://user-images.githubusercontent.com/29379759/121370414-b6e13300-c90a-11eb-9d22-b49795f184bd.png)


![image](https://user-images.githubusercontent.com/29379759/121370521-cb253000-c90a-11eb-8f91-20f568eced71.png)

**Comment by 4n4nd** [comment]
@chambridge looks like the current default storageclass isn't working for some reason, as a workaround could you please try using `ocs-storagecluster-cephfs` storageclass for now?

**Comment by chambridge** [comment]
Unfortunately, while postgres had a [custom resource field ](https://github.com/pulp/pulp-operator/blob/main/deploy/olm-catalog/pulp-operator/manifests/pulp.pulpproject.org_pulps_crd.yml#L105-L107)for this redis did not. I have the following [PR](https://github.com/pulp/pulp-operator/pull/154) out to add this capability into the operator.

**Comment by 4n4nd** [comment]
@chambridge the issue with the storageclass should be resolved and your PVCs should be provisioned.

**Comment by tumido** [comment]
Pulp is hitting the quota again (this is expected, since we're not familiar with the requirements). After a chat with @chambridge we decided to bump the namespace quota to 8 CPUs for now and see where that leads us.

**Comment by chambridge** [comment]
Pulp is up and running in the `thoth-pulp-experiments` namespace. I'm not sure if the problem from before was the NFS PVC problem or more likely the quota issue that you were hitting. Not knowing your base storage class was NFS and should support ReadWriteMany, I had went a head and used a Ceph ObjectBucket for storage and configured the S3 storage capability.

The CR for deployment looks like the following:
```
apiVersion: pulp.pulpproject.org/v1beta1
kind: Pulp
metadata:
  name: example-pulp
  namespace: thoth-pulp-experiments
spec:
  tag: latest
  storage_type: S3
  file_storage:
    size: 50Gi
  object_storage_s3_secret: pulp-object-storage
  ingress_type: Route
  route_tls_termination_mechanism: Edge
  route_host: pulp.apps.zero.massopen.cloud
  content:
    replicas: 1
    resource_requirements:
      requests:
        cpu: 150m
        memory: 256Mi
      limits:
        cpu: 500m
  worker:
    replicas: 1
    resource_requirements:
      requests:
        cpu: 150m
        memory: 
      limits:
        cpu: 500m
  web:
    replicas: 1
    resource_requirements:
      requests:
        cpu: 100m
        memory: 256Mi
      limits:
        cpu: 500m
  resource_manager:
    resource_requirements:
      requests:
        cpu: 100m
        memory: 256Mi
      limits:
        cpu: 500m
  postgres_resource_requirements:
    limits:
      cpu: 500m
  redis_resource_requirements:
    limits:
      cpu: 500m
```

You may note when looking at the CR that it currently says its failing. The operator is hitting an error when checking the Pulp instance's status endpoint. The problem should be resolved by the following [PR](https://github.com/pulp/pulp-operator/pull/155) which adds the service cluster local DNS look up, since the operator and pulp instance are in different namespaces.

I did validate it was running by uploading an Ansible collection thru the API:
![image](https://user-images.githubusercontent.com/29379759/121419090-97153380-c939-11eb-8b07-d6342533a542.png)

**Comment by fridex** [comment]
@chambridge thanks! 👍🏻

**Comment by fridex** [comment]
It looks like it is not possible to upload Python packages to pulp - the status endpoint shows pulp_python version 3.4.0.dev that is before version 3.4.0 with full twine upload support based on @gerrod3. The pulp_python==3.4.0 release was not propagated to the op1st environment. Can we update to pulp_python==3.4.0?

**Comment by chambridge** [comment]
I updated the Custom Resource with:
```
image_pull_policy: Always
```
This caused the operator to redeploy pulling the latest image from quay.io instead of using the latest tagged image that was within the local registry.

![image](https://user-images.githubusercontent.com/29379759/123338921-9ecaff80-d517-11eb-9daa-a3f478967357.png)

**Comment by fridex** [comment]
Thanks @chambridge. I've tried to upload a package, the upload finished successfully but the package is not seen on pulp (using admin creds):

```
$ twine upload micropipenv-1.1.0.tar.gz --repository-url https://pulp.operate-first.cloud/pypi/fpokorny-test/simple/
Uploading micropipenv-1.1.0.tar.gz
100%|█████████████████████| 55.2k/55.2k [00:02<00:00, 28.2kB/s]
$ echo $?
0
```

```
$ curl https://pulp.operate-first.cloud/pypi/fpokorny-test/
{"projects":0,"releases":0,"files":0}
```

I see `.dev` suffix in the pulp component version. Is it expected?

CC @@gerrod3

**Comment by ipanova** [comment]
@thank you @chambridge !
@fridex these are unreleased versions, you're basically working with master, should not you append `simple` to the endpoint?
https://docs.pulpproject.org/pulp_python/workflows/pypi.html?highlight=twine#upload-and-install-packages

**Comment by fridex** [comment]
> @thank you @chambridge !
> @fridex these are unreleased versions, you're basically working with master, should not you append `simple` to the endpoint?

Originally tried with just "`simple`" during upload - it resulted in an error:

```
$ twine upload ../micropipenv/dist/micropipenv-1.1.0.tar.gz --repository-url https://pulp.operate-first.cloud/pypi/fpokorny-test/simple
Uploading micropipenv-1.1.0.tar.gz
100%|█████████████████████| 55.2k/55.2k [00:01<00:00, 36.8kB/s]
Received "500: Internal Server Error" Package upload appears to have failed.  Retry 1 of 5
Uploading micropipenv-1.1.0.tar.gz
100%|█████████████████████| 55.2k/55.2k [00:01<00:00, 52.8kB/s]
Received "500: Internal Server Error" Package upload appears to have failed.  Retry 2 of 5
Uploading micropipenv-1.1.0.tar.gz
100%|█████████████████████| 55.2k/55.2k [00:00<00:00, 94.2kB/s]
Received "500: Internal Server Error" Package upload appears to have failed.  Retry 3 of 5
Uploading micropipenv-1.1.0.tar.gz
100%|█████████████████████| 55.2k/55.2k [00:00<00:00, 72.0kB/s]
Received "500: Internal Server Error" Package upload appears to have failed.  Retry 4 of 5
Uploading micropipenv-1.1.0.tar.gz
100%|█████████████████████| 55.2k/55.2k [00:01<00:00, 54.3kB/s]
Received "500: Internal Server Error" Package upload appears to have failed.  Retry 5 of 5
NOTE: Try --verbose to see response content.
HTTPError: 500 Server Error: Internal Server Error for url: https://pulp.operate-first.cloud/pypi/fpokorny-test/simple
```

The `--verbose` flag shows HTML with the internal error reported. I see the following logs in Pulp API pod:

```
RuntimeError: You called this URL via POST, but the URL doesn't end in a slash and you have APPEND_SLASH set. Django can't redirect to the slash URL while maintaining POST data. Change your form to point to pulp.operate-first.cloud/pypi/fpokorny-test/simple/ (note the trailing slash), or set APPEND_SLASH=False in your Django settings.
pulp [cdc4a1f705dd4e45938f53e4f4465ddb]: django.request:ERROR: Internal Server Error: /pypi/fpokorny-test/simple
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/utils/deprecation.py", line 93, in __call__
    response = self.process_request(request)
  File "/usr/local/lib/python3.9/site-packages/django/middleware/common.py", line 53, in process_request
    path = self.get_full_path_with_slash(request)
  File "/usr/local/lib/python3.9/site-packages/django/middleware/common.py", line 86, in get_full_path_with_slash
    raise RuntimeError(
RuntimeError: You called this URL via POST, but the URL doesn't end in a slash and you have APPEND_SLASH set. Django can't redirect to the slash URL while maintaining POST data. Change your form to point to pulp.operate-first.cloud/pypi/fpokorny-test/simple/ (note the trailing slash), or set APPEND_SLASH=False in your Django settings.
```

Based on the exception, tried with `simple/`. The upload succeeded (see log above) but no packages are listed: https://pulp.operate-first.cloud/pypi/fpokorny-test/simple/

Thanks for any additional pointers.

**Comment by ipanova** [comment]
I'm afraid @gerrod or @dralley should look into this

**Comment by fridex** [comment]
Thanks, Ina - debugging this with Gerrod. Thanks to @gerrod3 for pointers on how to debug this. Will report back any progress on this.

**Comment by fridex** [comment]
> Thanks, Ina - debugging this with Gerrod. Thanks to @gerrod3 for pointers on how to debug this. Will report back any progress on this.

We face the following issues:

* on the worker side - worker fails to finish upload (ceph/s3 related?):

```
pulp [d9fcb50ef55843a0930feba8feeba1de]: rq.worker:ERROR: Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/rq/worker.py", line 1013, in perform_job
    rv = job.perform()
  File "/usr/local/lib/python3.9/site-packages/rq/job.py", line 709, in perform
    self._result = self._execute()
  File "/usr/local/lib/python3.9/site-packages/rq/job.py", line 732, in _execute
    result = self.func(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/pulp_python/app/tasks/upload.py", line 43, in upload
    content_to_add = pre_check or create_content(artifact_sha256, filename)
  File "/usr/local/lib/python3.9/site-packages/pulp_python/app/tasks/upload.py", line 105, in create_content
    shutil.copy2(artifact.file.path, temp_path)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/fields/files.py", line 58, in path
    return self.storage.path(self.name)
  File "/usr/local/lib/python3.9/site-packages/django/core/files/storage.py", line 116, in path
    raise NotImplementedError("This backend doesn't support absolute paths.")
NotImplementedError: This backend doesn't support absolute paths.
```

* multiple uploads result in unique constraint issues:

```
2021-06-28 14:27:33.570 UTC [708432] ERROR:  duplicate key value violates unique constraint "core_artifact_sha256_key"
2021-06-28 14:27:33.570 UTC [708432] DETAIL:  Key (sha256)=(1e748895ba9d24cc3ecbe8004c22949f437d3ac7bb7db99fe66282fd72d9e038) already exists.
2021-06-28 14:27:33.570 UTC [708432] STATEMENT:  INSERT INTO "core_artifact" ("pulp_id", "pulp_created", "pulp_last_updated", "file", "size", "md5", "sha1", "sha224", "sha256", "sha384", "sha512") VALUES ('55a14f0c-0750-431d-9dea-40cf21f8d4a8'::uuid, '2021-06-28T14:27:33.570438+00:00'::timestamptz, '2021-06-28T14:27:33.570468+00:00'::timestamptz, 'artifact/1e/748895ba9d24cc3ecbe8004c22949f437d3ac7bb7db99fe66282fd72d9e038', 36370, NULL, NULL, '8bac2d46081e9f03290f4edbe4ed075862b38c431afdaa708cf8f3e2', '1e748895ba9d24cc3ecbe8004c22949f437d3ac7bb7db99fe66282fd72d9e038', '9f4e3d8f53f286701113a497dff4f9fb237c0ee40047d03433c4cfe765296bef2a61fb5034abd97913af7023bd352df0', 'b338ecf60cabab561598259729eedc67c6e7b5e638cdc482acca28844e995e6cd466b0dd2e35037122ea0c8fd892ad89006dab42089f294ecb136705e863704a')
```

@tumido would it be possible to give @gerrod3 (gubben) access to the `thoth-pulp-experiments` namespace to help with debugging? Thanks in advance. 👍🏻

**Comment by tumido** [comment]
@fridex done, @gerrod3 should have access now

**Comment by fridex** [comment]
The twine support should be fixed, thanks to @gerrod3 - now waiting for https://github.com/pulp/pulp-operator/pull/165 to get merged.

**Comment by fao89** [comment]
> The twine support should be fixed, thanks to @gerrod3 - now waiting for [pulp/pulp-operator#165](https://github.com/pulp/pulp-operator/pull/165) to get merged.

I just merged it!

**Comment by fridex** [comment]
> > The twine support should be fixed, thanks to @gerrod3 - now waiting for [pulp/pulp-operator#165](https://github.com/pulp/pulp-operator/pull/165) to get merged.
> 
> I just merged it!

Thanks, Fabricio. What is needed on the deployment side to apply the change on the Operate First instance? (I'm still getting task errors, I understand your fix is related)

**Comment by fao89** [comment]
you need to pull the images again, like this
https://github.com/operate-first/support/issues/176#issuecomment-867987933

**Comment by fridex** [comment]
Tried to reprovision the whole pulp instance after some experimental changes. It looks like the whole deployment was brought up with the operator. Unfortunatelly, when trying to create a pulp python repository, I'm getting the following error on API:

```
Internal Server Error: /pulp/api/v3/repositories/python/python/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedColumn: column "retained_versions" of relation "core_repository" does not exist
LINE 1: ...thoth', "description" = NULL, "next_version" = 0, "retained_...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/mixins.py", line 19, in create
    self.perform_create(serializer)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/mixins.py", line 24, in perform_create
    serializer.save()
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 205, in save
    self.instance = self.create(validated_data)
  File "/usr/local/lib/python3.9/site-packages/pulpcore/app/serializers/base.py", line 161, in create
    instance = super().create(validated_data)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 939, in create
    instance = ModelClass._default_manager.create(**validated_data)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 422, in create
    obj.save(force_insert=True, using=self.db)
  File "/usr/local/lib/python3.9/site-packages/pulpcore/app/models/repository.py", line 90, in save
    super().save(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/pulpcore/app/models/base.py", line 149, in save
    return super().save(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django_lifecycle/mixins.py", line 134, in save
    save(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 743, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 779, in save_base
    parent_inserted = self._save_parents(cls, using, update_fields)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 808, in _save_parents
    updated = self._save_table(
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 853, in _save_table
    updated = self._do_update(base_qs, using, pk_val, values, update_fields,
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 903, in _do_update
    return filtered._update(values) > 0
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 760, in _update
    return query.get_compiler(self.db).execute_sql(CURSOR)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1471, in execute_sql
    cursor = super().execute_sql(result_type)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1142, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 99, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column "retained_versions" of relation "core_repository" does not exist
LINE 1: ...thoth', "description" = NULL, "next_version" = 0, "retained_...
                                                             ^

pulp [1f49704d05e94a3fba0980b84bd254ee]: django.request:ERROR: Internal Server Error: /pulp/api/v3/repositories/python/python/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedColumn: column "retained_versions" of relation "core_repository" does not exist
LINE 1: ...thoth', "description" = NULL, "next_version" = 0, "retained_...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/mixins.py", line 19, in create
    self.perform_create(serializer)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/mixins.py", line 24, in perform_create
    serializer.save()
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 205, in save
    self.instance = self.create(validated_data)
  File "/usr/local/lib/python3.9/site-packages/pulpcore/app/serializers/base.py", line 161, in create
    instance = super().create(validated_data)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 939, in create
    instance = ModelClass._default_manager.create(**validated_data)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 422, in create
    obj.save(force_insert=True, using=self.db)
  File "/usr/local/lib/python3.9/site-packages/pulpcore/app/models/repository.py", line 90, in save
    super().save(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/pulpcore/app/models/base.py", line 149, in save
    return super().save(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django_lifecycle/mixins.py", line 134, in save
    save(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 743, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 779, in save_base
    parent_inserted = self._save_parents(cls, using, update_fields)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 808, in _save_parents
    updated = self._save_table(
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 853, in _save_table
    updated = self._do_update(base_qs, using, pk_val, values, update_fields,
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 903, in _do_update
    return filtered._update(values) > 0
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 760, in _update
    return query.get_compiler(self.db).execute_sql(CURSOR)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1471, in execute_sql
    cursor = super().execute_sql(result_type)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1142, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 99, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column "retained_versions" of relation "core_repository" does not exist
LINE 1: ...thoth', "description" = NULL, "next_version" = 0, "retained_...
                                                             ^

pulp [d073c5b6fcbd47a5997cc442f1ff3863]: pulpcore.app.views.status:ERROR: Connection to Redis failed during status check!
pulp [b6e1dc5d4d5645eab6202b9576fe5615]: pulpcore.app.views.status:ERROR: Connection to Redis failed during status check!
```

Thanks for any pointers how to fix this.

**Comment by fridex** [comment]
Yesterday we had a debugging session with @gerrod3 and @fao89. The issue I reported above does not look relevant anymore after reprovisioning the pulp instance. twine uploads work now, we are able to upload built packages. However, the download is not operational. @gerrod3 suggested adjusting `PULP_CONTENT_ORIGIN` ([see docs](https://docs.pulpproject.org/pulpcore/configuration/settings.html#content-origin)) environment variable for api service so it correctly redirects to the content service. We are experiencing limitations in resources allocated for the namespace - see https://github.com/operate-first/support/issues/321 request.

**Comment by ipanova** [comment]
@fridex thank you for the summary, that was helpful to understand where is the problem and that there is progress around it.

**Comment by fao89** [comment]
@tumido I've been seeing this at the operator logs:
Actually, I found another error, pulp-operator keeps reconciling with this error:

```
TASK [pulp-api : pulp-file-storage persistent volume claim] ********************\r
\u001b[1;30mtask path: /opt/ansible/roles/pulp-api/tasks/main.yml:21\u001b[0m
\u001b[0;31mfailed: [localhost] (item=pulp-file-storage) => {\"ansible_loop_var\": \"item\", \"changed\": false, \"error\": 403, \"item\": \"pulp-file-storage\", \"msg\": \"Failed to patch object: b'{\\\"kind\\\":\\\"Status\\\",\\\"apiVersion\\\":\\\"v1\\\",\\\"metadata\\\":{},\\\"status\\\":\\\"Failure\\\",\\\"message\\\":\\\"persistentvolumeclaims \\\\\\\\\\\"example-pulp-file-storage\\\\\\\\\\\" is forbidden: exceeded quota: thoth-pulp-experiments-custom, requested: requests.storage=50Gi, used: requests.storage=59Gi, limited: requests.storage=60Gi\\\",\\\"reason\\\":\\\"Forbidden\\\",\\\"details\\\":{\\\"name\\\":\\\"example-pulp-file-storage\\\",\\\"kind\\\":\\\"persistentvolumeclaims\\\"},\\\"code\\\":403}\\\
'\", \"reason\": \"Forbidden\", \"status\": 403}
```
I suspect it may be related to NFS settings, could you please take a look?

**Comment by tumido** [comment]
Hey, pulp was migrated to a new cluster - Smaug, because Zero wen't down. Is there anything to be done for this issue or can we consider the onboarding complete as of now?

I think we should maybe think about closing this issue if the initial setup is done and rather resort to opening new issues if any problems show up.

**Comment by fridex** [comment]
Thanks for the work and migration 👍🏻

Checking https://pulp.operate-first.cloud, the index is not yet reachable.

> I think we should maybe think about closing this issue if the initial setup is done and rather resort to opening new issues if any problems show up.

Sounds like a good idea. Let's finish deployment and have mvp up so we can close this and report any issues along the way.

Thanks again.

**Comment by harshad16** [comment]
Facing the issue on getting the pods on same node, so created the issue on upstream with pulp-operator
https://github.com/pulp/pulp-operator/issues/289

**Comment by harshad16** [comment]
Due release of pulp-operator: v0.6.1, the readwrite issue is fixed with the node selection. 

newer issue causing the hinderance: 
```
  File "/usr/local/lib64/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
django.db.utils.OperationalError: FATAL:  password authentication failed for user "pulp"
```

**Comment by harshad16** [comment]
Opened upstream issue : https://github.com/pulp/pulp-operator/issues/308

**Comment by chambridge** [comment]
I've hit a similar issue to this in the past during development if I didn't delete the persistent volume claim for postgres (which isn't deleted when you delete your pulp custom resource) and then I create a new instance with a new postgres deploy on top of the same pvc. The generated password (secret) for the new pulp instance's postgres doesn't match what has been initialized in the previous and the db user can't access. Just a though, in case that was the issue.

**Comment by harshad16** [comment]
thank you , that was the issue
deleting the old PVC helped.

**Comment by harshad16** [comment]
Old the setup is running. 
The api is having trouble connecting the redis db:
```
pulp [eff4343985714d8bb0b891b0b6e3e5ce]: pulpcore.app.views.status:ERROR: Connection to Redis failed during status check!
```
status can be seen here: https://pulp.operate-first.cloud/pulp/api/v3/status/

**Comment by harshad16** [comment]
The pulp instance is working and running on the operate-first cluster.
https://console-openshift-console.apps.smaug.na.operate-first.cloud/k8s/ns/opf-pulp/pods

The status can be seen here: https://pulp.operate-first.cloud/pulp/api/v3/status/
Example upload: https://pulp.operate-first.cloud/pypi/hello-world/simple/aicoe-hello-world/

Admin can create the index:
```
pulp python repository create --name hello-world
pulp python distribution create --base-path hello-world --name hello-world --repository hello-world
```
and upload the package via twine:
```
twine upload dist/* --repository-url="https://pulp.operate-first.cloud/pypi/hello-world/simple/"
```

**Comment by fridex** [comment]
I think this can be closed and we can track issues separately as discussed above.

Thanks a lot to all for delivering this solution. 👏🏻 

/close
